Additional border_modes for _interpolate.

hjweide opened the pull request:
Closes #689.  Add additional border modes for _interpolate to be consistent with scipy.

f0k commented on lasagne/layers/special.py:
```diff
@@ -633,10 +649,12 @@ class TPSTransformerLayer(MergeLayer):
     def __init__(self, incoming, localization_network, downsample_factor=1,
-                 control_points=16, precompute_grid='auto', **kwargs):
+                 border_mode='nearest', control_points=16,
+                 precompute_grid='auto', **kwargs):
```
Please add the new argument in the end, to keep compatibility with existing code that uses positional arguments.

f0k commented on lasagne/layers/special.py:
```diff
@@ -452,23 +454,37 @@ def _transform_affine(theta, input, downsample_factor):
+        x = T.clip(x, -1, 1)
+        y = T.clip(y, -1, 1)
+    # 0.9 1.0 1.1 -> 0.9 1.0 0.9
+    elif border_mode == 'reflect':
```
Looking at http://stackoverflow.com/a/22670830, I think this should be `'mirror'` instead. `reflect` is more difficult to implement since it depends on the discrete pixel positions, we won't need this.

f0k commented on lasagne/layers/special.py:
```diff
@@ -452,23 +454,37 @@ def _transform_affine(theta, input, downsample_factor):
+        xa = T.maximum(x, -(2 + x))
+        ya = T.maximum(y, -(2 + y))
+        x = T.minimum(xa, 2 - xa)
+        y = T.minimum(ya, 2 - ya)
```
Almost! It needs to be periodic just like `wrap`. E.g., when you go further, 2.9 should be mapped to -.9 (which works), 3.0 should be mapped to -1 (which works), and 3.1 should be mapped to -.9 again (mapped to -1.1 now).

hjweide commented:
Getting ```mirror``` right was a little trickier than I first thought, but I think I managed to make it consistent with scipy.  I've also changed the order of the arguments and added docstrings.  If those all look good to you I can add some tests too.

f0k commented on lasagne/layers/special.py:
```diff
@@ -376,6 +376,12 @@ class TransformerLayer(MergeLayer):
+        'nearest', points outside the grid are clipped to the boundary. If
+        'mirror', points are mirrored across the boundary. If 'wrap', points
+        wrap around to the other side of the grid.  See
+        http://stackoverflow.com/q/22669252/22670830#22670830 for details.
```
Not sure if we want to refer to stackoverflow there... but if we do, we could use a shorter link: `See http://stackoverflow.com/a/22670830 for an illustration.`

f0k commented on lasagne/layers/special.py:
```diff
@@ -452,23 +460,37 @@ def _transform_affine(theta, input, downsample_factor):
+        xa = T.mod(x + 1, 4) - 1
+        ya = T.mod(y + 1, 4) - 1
+        x = T.minimum(xa, 2 - xa)
+        y = T.minimum(ya, 2 - ya)
```
Perfect! 👍

hjweide commented:
Any suggestion how to test the ```_interpolate``` function?  It returns the image itself, so to test it we'd need to reproduce the effect of the border modes on the input image.

f0k commented:
> It returns the image itself, so to test it we'd need to reproduce the effect of the border modes on the input image.

Hmm, that would basically require to reimplement `_interpolate` in numpy. Maybe it's easier to test the TransformerLayer instead, with a fixed transform so the output image is very easy to compute in numpy. Something like an input of [0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1], horizontally squashed to 25% (but still filling 16 pixels), thresholded to make it binary again, and then we can check if the border mode was respected (each border mode will result in another simple pattern). Does that sound reasonable?

hjweide commented:
Thanks for the suggestion.  I've added tests to get the coverage back up to 100%.

f0k commented:
> I've added tests to get the coverage back up to 100%.

Yay! Nice and clean. I'm surprised the patterns become a bit irregular, but we probably can't do anything about this. Must be some unfortunate rounding.

As a sanity check, I tried an MNIST digit (loaded as single-image batch `x`):
```python
trans = np.array([[4, 0, 0, 0, 2, 0]], dtype=np.float32)
for border_mode in 'nearest', 'mirror', 'wrap':
    plt.figure()
    plt.imshow(lasagne.layers.TransformerLayer(x.shape, trans.shape, border_mode=border_mode, downsample_factor=(.5, .25)).get_output_for((T.constant(x), T.constant(trans))).eval()[0,0].T, cmap='gray', interpolation='none')
```
Looks nice (in order: nearest, mirror, wrap):
![nearest](https://cloud.githubusercontent.com/assets/629706/25450918/c1fe66e6-2ac0-11e7-8998-609e72bedcb4.png) ![mirror](https://cloud.githubusercontent.com/assets/629706/25450928/c71c7028-2ac0-11e7-92e7-6837a17ff59d.png) ![wrap](https://cloud.githubusercontent.com/assets/629706/25450933/c94d8eae-2ac0-11e7-8923-213e21220fed.png)

Again, upon close inspection, the mirrored/wrapped versions are not exactly identical to the original in the center (and also the version in the center is not exactly identical to the untransformed digit), but that's not the business of this PR.

Very good job! :+1: Will merge tomorrow if there are no further comments.

f0k commented:
Merged, thanks again for your great work!

hjweide commented:
Happy to help!  I think setting pixels outside the grid to a constant value would be a useful border mode to have, but I'm not sure how that would be implemented.  Perhaps I'll get back to it in the future.

f0k commented:
> I think setting pixels outside the grid to a constant value would be a useful border mode to have, but I'm not sure how that would be implemented.

I guess you just wouldn't clip the coordinates, and then use a `T.switch` when fetching the pixel values to replace out-of-image values with a given constant value here: https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/special.py#L527-L530

f0k commented:
Hmm, and looking at the code again, wouldn't we have to respect the border mode here as well? https://github.com/Lasagne/Lasagne/blob/master/lasagne/layers/special.py#L508-L509

hjweide commented:
Indeed you're right, no idea how I missed that.  I've implemented what I think is the correct fix, and will be pushing it momentarily.  This also seems to take care of the irregular patterns we were seeing.

f0k commented:
> Indeed you're right, no idea how I missed that.

Well, I also missed it, your solution was very convincing :)

> Turns out getting mirror right is a little tricky. I'll get back to it soon.

Thank you for taking this! I think the correct solution would be to change the `x0 = T.cast(x0_f)` and the other three conversions to respect the border mode, and I think you can use the code you have, just using `width_f` or `height_f` instead of `1` (and `4 * width_f` instead of `4`). Placing the border mode code there will also make it easier to change the grid domain as suggested in #800 to fix #782.

hjweide commented:
My first attempt was just to replace the previous width/height of 2 (from -1...1) with width_f/height_f as appropriate, but that leads to strange results for the mirror case.
```
x0a = T.mod(x0_f + width_f / 2, 2 * width_f) - width_f / 2
y0a = T.mod(y0_f + height_f / 2, 2 * height_f) - height_f / 2
x0 = T.minimum(x0a, width_f - x0a)
y0 = T.minimum(y0a, height_f - y0a)

x1a = T.mod(x1_f + width_f / 2, 2 * width_f) - width_f / 2
y1a = T.mod(y1_f + height_f / 2, 2 * height_f) - height_f / 2
x1 = T.minimum(x1a, width_f - x1a)
y1 = T.minimum(y1a, height_f - y1a)
```
Not sure if I'm missing something obvious but I may need to redo the calculations.
![mirror](https://cloud.githubusercontent.com/assets/6610704/25558294/b69e68b8-2cf1-11e7-869a-c2ac3bb5e9ba.png)

f0k commented:
You're right, the formulas need to change a bit. I have the following:
```python
    # obtain indices of the 2x2 pixel neighborhood surrounding the coordinates;
    # we need those in floatX for interpolation and in int64 for indexing.
    x0_f = T.floor(x)
    y0_f = T.floor(y)
    x1_f = x0_f + 1
    y1_f = y0_f + 1

    # for indexing, we need to take care of the border mode for outside pixels.
    if border_mode == 'nearest':
        x0 = T.clip(T.cast(x0_f, 'int64'), 0, width - 1)
        x1 = T.clip(T.cast(x1_f, 'int64'), 0, width - 1)
        y0 = T.clip(T.cast(y0_f, 'int64'), 0, height - 1)
        y1 = T.clip(T.cast(y1_f, 'int64'), 0, height - 1)
    elif border_mode == 'mirror':
        xa0 = T.mod(x0_f, 2 * width_f)
        xa1 = T.mod(x1_f, 2 * width_f)
        ya0 = T.mod(y0_f, 2 * height_f)
        ya1 = T.mod(y1_f, 2 * height_f)
        x0 = T.cast(T.minimum(xa0, 2 * width_f - 1 - xa0), 'int64')
        x1 = T.cast(T.minimum(xa1, 2 * width_f - 1 - xa1), 'int64')
        y0 = T.cast(T.minimum(ya0, 2 * height_f - 1 - ya0), 'int64')
        y1 = T.cast(T.minimum(ya1, 2 * height_f - 1 - ya1), 'int64')
    elif border_mode == 'wrap':
        x0 = T.cast(T.mod(x0_f, width_f), 'int64')
        x1 = T.cast(T.mod(x1_f, width_f), 'int64')
        y0 = T.cast(T.mod(y0_f, height_f), 'int64')
        y1 = T.cast(T.mod(y1_f, height_f), 'int64')
    else:
        raise ValueError("border_mode must be one of "
                         "'nearest', 'mirror', 'wrap'")
```
Not 100% sure about the `-1`.

hjweide commented:
Previously the grid ended at ```1```, so we mirrored about ```2``` before taking the minimum.  Now the grid ends at ```width_f - 1```, so does that mean we should mirror about ```2 * (width_f - 1)```?

f0k commented:
> Now the grid ends at `width_f - 1`, so does that mean we should mirror about `2 * (width_f - 1)`?

Well, there's a clear definition for the `mirror` mode at http://stackoverflow.com/a/22670830. My solution above might be the `reflect` mode instead, I only briefly tried it with a picture of a parrot, which was nice to look at, but not the best choice for scrutiny ;)
Let's just go through the formula for a toy image of three pixels. We would want the grid positions `-1, 0, 1` to be mapped to `0, 1, 2` (since the grid spans from the left corner of the first pixel (0) to the left corner of the third pixel (2)), this is catered for by the conversion to `x`. Grid positions `2, 3, 4, 5, 6` will be mapped to pixel positions `3, 4, 5, 6, 7`. We want the border mode code for `mirror` to map this to `1, 0, 1, 2, 1` (`reflect` would be `2, 1, 0, 0, 1` instead).
My solution above was `min(x mod 2*width, 2*width - 1 - x)`. With `width = 3` and `x = [3, 4, 5, 6, 7]`, this would give `[2, 1, 0, -1, -2]`. So that's wrong, the second part needs a `mod` as well. With `min(x mod 6, (5 - x) mod 6)` we'll have `reflect`:
```
>>> x = np.arange(12)
>>> np.minimum(x % 6, (5 - x) % 6)
array([0, 1, 2, 2, 1, 0, 0, 1, 2, 2, 1, 0])
```
The correct solution for `mirror` would be:
```
>>> np.minimum(x % 4, (4 - x) % 4)
array([0, 1, 2, 1, 0, 1, 2, 1, 0, 1, 2, 1])
```
Do you want to continue with this? (I'd suggest to not include `reflect`, as before, even if the two are both easy to implement now that we're in pixel space. I think `reflect` is usually not what you want, and including both makes it confusing for users.)
